Source/WebGPU/WGSL/AST/ASTWorkgroupSizeAttribute.h

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "ASTNode.h"
+
+#include <wtf/UniqueRef.h>
+#include <wtf/UniqueRefVector.h>
+#include <wtf/Vector.h>
+
+namespace WGSL::AST {
+
+class WorkgroupSizeAttribute final : public Attribute {
```

Comment:
2 questions:
1. Why is this its own node rather than simply being an enum field in a `class Attribute`?
2. Attributes are not recursive, so why make them AST nodes in the first place? Why not describe them via an elaborate `Variant` typedef instead of having them be heap-allocated and traversed by the Visitor?

Context: I made Attributes be AST nodes in WHLSL and I ended up regretting it - it cost more in implementation lines of code than it gained. None of the passes actually ended up using it without _also_ using the context of its owner. And if you need the context of the owner in order to do anything useful with an attribute, might as well just have the owner's visitor function just inspect any present attributes directly, rather than asking the Visitor to recurse to them.

Reply:
I'm mostly onboard, not sure about variants, but I agree they don't necessarily need to be traversed. I think I want to refactor a lot of this, but I want to sync with @djg before any drastic changes, so for now I'm just sticking to the current style and adding just enough to it to unblock myself.